### ISSUE TITLE
Enable pre-training by initializing new model with previously trained weights

### ISSUE THREAD

**Issue by rak5216** [mit-quest/necstlab-damage-segmentation#32]

**Comment by rak5216** [comment]
@Josh-Joseph do you think this will be a quick task for the next issue to tackle? Perhaps on the command line for training we can give a new parameter like `--pre-trained-weights <model_ID>`, with default being no pre-training when the parameter is excluded. Then it just loads in this model and starts epoch 1 with that.

**Comment by CarolinaFurtado** [comment]
Branch: **issue_32_init_pretrained_weigths**

Modified **train_segmetation_model.py:** 
- added a **new arg**: model_id_start, default to none
- if model_id_start not none (a pretrained model is used to initialize the weights): 
      - copy model locally 
      - compiled_model.weights_to_load = Path(local_model_dir, model_id_start, "model.hdf5").as_posix()

It runs, both with and without the model_id_start argument. Not sure if is doing the right thing yet. 

Still need to modify the metadata file to say where the initial weights come from.

**Comment by CarolinaFurtado** [comment]
Added the pretrained model to the metadata file: 

```
created_datetime: 20201019T164941Z
dataset_config:
  class_annotation_mapping:
    class_0_annotation_GVs: [100]
    class_1_annotation_GVs: [175]
    class_2_annotation_GVs: [250]
  dataset_split:
    test: [8bit_AS4_S2_P1_L6_2560_1750_2160]
    train: [THIN_REF_S2_P1_L3_2496_1563_2159]
    validation: [THIN_CNT_S2_P1_L4_2334_1578_2159]
  image_cropping:
    min_num_class_pos_px: {class_0_pos_px: 5, class_1_pos_px: 5, class_2_pos_px: 5}
    num_neg_per_class: 1
    num_per_image: None
    num_pos_per_class: 1
    type: class
  stack_downsampling: {num_skip_beg_slices: 50, num_skip_end_slices: 50, number_of_images: 100,
    type: linear}
  target_size: &id001 [512, 512]
elapsed_minutes: 8.6
gcp_bucket: gs://necstlab-sandbox
git_hash: 692429e0290544b2652339fb4e8e99050054e591
global_threshold_for_metrics: 0.5
num_classes: 3
original_config_filename: configs/config_sandbox/tf_version_debug//train-small-3class_tfversion_debug.yaml
**pre_trained_model_config:
  batch_size: 16
  data_augmentation: {random_90-degree_rotations: false}
  dataset_id: dataset-small-3class_tfversion_debug
  epochs: 10
  loss: cross_entropy
  model_id_prefix: segmentation-model-small-3class_tfversion_debug
  optimizer: adam
  segmentation_model:
    model_name: Unet
    model_parameters: {backbone_name: vgg16, encoder_weights: null}
  test_data_shuffle_seed: 123456
  training_data_shuffle_seed: 1234
  validation_data_shuffle_seed: 12345
pre_trained_model_id: segmentation-model-small-3class_tfversion_debug_20201016T164701Z**
target_size: *id001
```

**Comment by CarolinaFurtado** [comment]
If we train a model: 
`python3 train_segmentation_model.py --gcp-bucket gs://necstlab-sandbox --config-file configs/config_sandbox/tf_version_debug//train-small-3class_tfversion_debug.yaml `

 and then train a model starting with the first models pretrained weights: 

`python3 train_segmentation_model.py --gcp-bucket gs://necstlab-sandbox --config-file configs/config_sandbox/tf_version_debug//train-small-3class_tfversion_debug.yaml --pre-trained-model-id segmentation-model-small-3class_tfversion_debug_20201019T174848Z
`

This is what we get: - the second model continuous training with the loss of the previously trained model: 
![image](https://user-images.githubusercontent.com/43469359/96499003-cc3e0b80-121a-11eb-93a2-f3dcc20a06c2.png)

**Comment by CarolinaFurtado** [comment]
If we train two models from the first trained models: 
`python3 train_segmentation_model.py --gcp-bucket gs://necstlab-sandbox --config-file configs/config_sandbox/tf_version_debug//train-small-3class_tfversion_debug.yaml --pre-trained-model-id segmentation-model-small-3class_tfversion_debug_20201019T174848Z
`
![image](https://user-images.githubusercontent.com/43469359/96499015-d3fdb000-121a-11eb-93e0-10c7580b91a8.png)

we obtain very similar loss evolutions. The starting point is the same, evolution is very similar but not exactly the same. All the possible (known) source of randomness is eliminated,   `training_data_shuffle_seed`, `validation_data_shuffle_seed` and  `test_data_shuffle_seed` and `random_90-degree_rotations`: 

```
train_config:
  model_id_prefix: 'segmentation-model-small-3class_tfversion_debug'
  dataset_id: 'dataset-small-3class_tfversion_debug'
  segmentation_model:
    model_name: 'Unet'
    model_parameters:
      backbone_name: 'vgg16'
      encoder_weights: Null   # Null (random initialization) or 'imagenet' (pre-training on ImageNet)
  loss: 'cross_entropy'
  optimizer: 'adam'
  batch_size: 16    # p100
  epochs: 10
  training_data_shuffle_seed: 1234
  validation_data_shuffle_seed: 12345
  test_data_shuffle_seed: 123456
  data_augmentation:
    random_90-degree_rotations: False
```

**Comment by CarolinaFurtado** [comment]
To test if the seeds are working or not, we set the `shuffle `to `False `in the `train_segmentation_model.py`:

```
    train_generator = ImagesAndMasksGenerator(
        Path(local_dataset_dir, train_config['dataset_id'], 'train').as_posix(),
        rescale=1. / 255,
        target_size=target_size,
        batch_size=batch_size,
        shuffle=False,
        random_rotation=train_config['data_augmentation']['random_90-degree_rotations'],
        seed=None if 'training_data_shuffle_seed' not in train_config else train_config['training_data_shuffle_seed'])
```

This would eliminate the source of randomness for sure. However, we still get slightly different loss evolutions when training 2 models with this configuration: 

`python3 train_segmentation_model.py --gcp-bucket gs://necstlab-sandbox --config-file configs/config_sandbox/tf_version_debug//train-small-3class_tfversion_debug.yaml --pre-trained-model-id segmentation-model-small-3class_tfversion_debug_20201019T174848Z
`

![image](https://user-images.githubusercontent.com/43469359/96499932-1378cc00-121c-11eb-8b17-f5217915339e.png)


It is my understanding that this means that tensorflow/ damage segmentation models have some intrinsic source of randomness that we do not control. However, it does not seem significant

**Comment by CarolinaFurtado** [comment]
**Josh's comment:** 
ah so I wonder if this is randomness then of the optimizer -- so we'd also need to make sure the optimizer is seeded. I don't think(?) you can do that directly from keras but you can set tensorflow's random seed https://www.tensorflow.org/api_docs/python/tf/random/set_seed (so this could be done before training starts)

https://machinelearningmastery.com/reproducible-results-neural-networks-keras/

**Reed's comment:** 
also, try putting this at the very top of the train script
```
from numpy.random import seed
seed(1)
from tensorflow import set_random_seed
set_random_seed(2)

```
if that doesn't work, then i think we will also need to make the GPU solve in a deterministic way (no issue here if using CPU tho). so then last thing will be: https://machinelearningmastery.com/reproducible-results-neural-networks-keras/

i guess these seeded modules at the top of train script shouldn't conflict with those in the dataset generators which are instantiated

**Comment by CarolinaFurtado** [comment]
Even after seeding `tensorflow.random`, `numpy.random` and `random `, the results were slightly different. 

We could obtain repeatable results running in CPU, which is deterministic. 

![image](https://user-images.githubusercontent.com/43469359/96741136-822a6680-138f-11eb-8d7d-603e43045aaa.png)

Now we can obtain repeatable results and have confirmed that training from pretrained weights is working. 

Now we'll see if the tf 2.1 and 2.3 results are the same: 
upgrade software versions #75

**Comment by CarolinaFurtado** [comment]
![image](https://user-images.githubusercontent.com/43469359/97305651-c60be880-1833-11eb-9f91-0fa6d0410a75.png)

the results with tf2.3 are also repeatable (using cpu) @rak5216 @Josh-Joseph
